### PR TITLE
reshape and document the dht-cache

### PR DESCRIPTION
**Pull request by lu-zero**
- Reduce the public API
- Document it

### CODE REVIEW

File: dht-cache/src/cache/local.rs [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+            storage.store(elem).await;
+        }
+
+        cache.put(elem);
+    }
+
+    /// Try to insert the element in the cache
+    ///
+    /// Return Err(()) if the element to insert is older than the one in the cache
+    pub async fn try_put(&self, elem: &DomoCacheElement) -> Result<(), ()> {
```
Comment (dodomorandi) [inline dht-cache/src/cache/local.rs:96]: I don't like this approach, IMHO it is better to return a `bool` at this point and mark the function as `#[must_use]`. It should also make the implementation a bit more clear.

File: dht-cache/src/cache/local.rs [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+    cache: LocalCache,
+    topic: String,
+    uuid: Option<String>,
+}
+
+impl Query {
+    /// Create a new query over a local cache
+    pub fn new(topic: &str, cache: LocalCache) -> Self {
+        Self {
+            topic: topic.to_owned(),
```
Comment (dodomorandi) [inline dht-cache/src/cache/local.rs:162]:
```suggestion
    pub fn new(topic: impl Into<String>, cache: LocalCache) -> Self {
        Self {
            topic: topic.into(),
```

File: dht-cache/src/cache/local.rs [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+            cache,
+            uuid: None,
+        }
+    }
+    /// Look up for a specific uuid
+    pub fn with_uuid(mut self, uuid: &str) -> Self {
+        self.uuid = Some(uuid.to_owned());
```
Comment (dodomorandi) [inline dht-cache/src/cache/local.rs:169]:
```suggestion
    pub fn with_uuid(mut self, uuid: impl Into<String>) -> Self {
        self.uuid = Some(uuid.into());
```

File: dht-cache/src/cache.rs [hunk 4]
```diff
@@ -0,0 +1,462 @@
...
+                    let der = pem_rfc7468::decode_vec(&pem)?;
+                    der.1
+                }
+                Err(e) if e.kind() == std::io::ErrorKind::NotFound => {
+                    // Generate a new key and put it into the file at the given path
+                    let (pem, der) = generate_rsa_key();
+                    std::fs::write(pk_path, pem)?;
+                    der
+                }
+                Err(e) => Err(e)?,
```
Comment (dodomorandi) [inline dht-cache/src/cache.rs:60]:
```suggestion
                Err(e) => return Err(e.into()),
```

File: dht-cache/src/cache.rs [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+    pub last_seen: u128,
+}
+
+impl Cache {
+    /// Send a volatile message
+    ///
+    /// Volatile messages are unstructured and do not persist in the DHT.
+    pub fn send(&self, value: &Value) -> Result<(), Error> {
+        self.cmd
+            .send(Command::Broadcast(value.to_owned()))
```
Comment (dodomorandi) [inline dht-cache/src/cache.rs:104]:
```suggestion
    pub fn send(&self, value: Value) -> Result<(), Error> {
        self.cmd
            .send(Command::Broadcast(value))
```

File: dht-cache/src/cache.rs [hunk 6]
```diff
@@ -0,0 +1,462 @@
...
+    }
+
+    /// Persist a value within the DHT
+    ///
+    /// It is identified by the topic and uuid value
+    pub async fn put(&self, topic: &str, uuid: &str, value: &Value) -> Result<(), Error> {
+        let elem = DomoCacheElement {
+            topic_name: topic.to_string(),
+            topic_uuid: uuid.to_string(),
+            value: value.to_owned(),
```
Comment (dodomorandi) [inline dht-cache/src/cache.rs:117]:
```suggestion
    pub async fn put(&self, topic: impl Into<String>, uuid: impl Into<String>, value: Value) -> Result<(), Error> {
        let elem = DomoCacheElement {
            topic_name: topic.into(),
            topic_uuid: uuid.into(),
            value,
```

File: dht-cache/src/cache.rs [hunk 7]
```diff
@@ -0,0 +1,462 @@
...
+    }
+
+    /// Delete a value within the DHT
+    ///
+    /// It inserts the deletion entry and the entry value will be marked as deleted and removed
+    /// from the stored cache.
+    pub async fn del(&self, topic: &str, uuid: &str) -> Result<(), Error> {
+        let elem = DomoCacheElement {
+            topic_name: topic.to_string(),
+            topic_uuid: uuid.to_string(),
```
Comment (dodomorandi) [inline dht-cache/src/cache.rs:139]:
```suggestion
    pub async fn del(&self, topic: impl Into<String>, uuid: impl Into<String>) -> Result<(), Error> {
        let elem = DomoCacheElement {
            topic_name: topic.into(),
            topic_uuid: uuid.into(),
```

File: dht-cache/src/cache.rs [hunk 8]
```diff
@@ -0,0 +1,462 @@
...
+        peer_id: local_peer_id.clone(),
+    };
+
+    let stream = UnboundedReceiverStream::new(r);
+
+    let local_read = local.clone();
+    let cmd_update = cmd.clone();
+    let peer_id = local_peer_id.clone();
+
+    tokio::task::spawn(async move {
```
Comment (dodomorandi) [inline dht-cache/src/cache.rs:255]: I think that this could be transformed into a `spawn_local` without any issue, leading to the possibility of removing the `unsafe impl std::marker::Sync for InnerCache {}`

File: dht-cache/src/cache.rs [hunk 9]
```diff
@@ -0,0 +1,462 @@
...
+                            .await
+                            .mem
+                            .values()
+                            .flat_map(|topic| topic.values())
+                            .for_each(|elem| {
+                                let mut elem = elem.to_owned();
+                                log::debug!("resending {}", elem.topic_uuid);
+                                elem.republication_timestamp = utils::get_epoch_ms();
+                                cmd.send(Command::Publish(serde_json::to_value(&elem).unwrap()))
+                                    .unwrap();
```
Comment (dodomorandi) [inline dht-cache/src/cache.rs:321]:
```suggestion
                               
                                // This cannot fail because `cmd` is the sender part of the
                                // `stream` we are currently reading. In practice, we are
                                // queueing the commands in order to read them later.
                                cmd.send(Command::Publish(serde_json::to_value(&elem).unwrap()))
                                    .unwrap();
```

File: dht-cache/src/cache.rs [hunk 10]
```diff
@@ -0,0 +1,462 @@
...
+        let (a_c, a_ev) = cache_channel(a_local_cache, a, 100);
+        let (b_c, b_ev) = cache_channel(b_local_cache, b, 100);
+        let (c_c, c_ev) = cache_channel(c_local_cache, c, 100);
+
+        let mut expected_peers = HashSet::new();
+        expected_peers.insert(a_c.peer_id.clone());
+        expected_peers.insert(b_c.peer_id.clone());
+        expected_peers.insert(c_c.peer_id.clone());
+
+        tokio::task::spawn(async move {
```
Comment (dodomorandi) [inline dht-cache/src/cache.rs:405]: I think we should avoid spawning tasks in tests, and "just" joining them. Spawning tasks in the runtime could make edge cases work by chance that otherwise, if we are lucky enough, could more easily deadlock in single thread and be detected by tests.

File: dht-cache/src/cache.rs [hunk 11]
```diff
@@ -0,0 +1,462 @@
...
+                match ev {
+                    DomoEvent::PersistentData(data) => {
+                        log::debug!("{node}: Got data {data:?}");
+                    }
+                    _ => {
+                        log::debug!("{node}: Other {ev:?}");
+                    }
+                }
+            }
+        });
```
Comment (dodomorandi) [inline dht-cache/src/cache.rs:435]: This is not cancel safe, and this is the reason we should avoid using `select!` when possible. This is a possible approach using `futures-concurrency`:

```rust
            (
                a_ev.map(|v| ("a", v)),
                b_ev.map(|v| ("b", v)),
                c_ev.map(|v| ("c", v)),
            )
                .merge()
                .for_each(|(node, ev)| async move {
                    match ev {
                        DomoEvent::PersistentData(data) => {
                            log::debug!("{node}: Got data {data:?}");
                        }
                        _ => {
                            log::debug!("{node}: Other {ev:?}");
                        }
                    }
                })
                .await;
```

**Review by dodomorandi — CHANGES_REQUESTED** [review]
I still need to finish reviewing, I have two major concerns about use of `unsafe` and `select!` other things are minor.

File: dht-cache/src/cache/local.rs (hunk 1 above)
Comment (lu-zero) [reply dht-cache/src/cache/local.rs:96]: I was on the fence between the two, given the usage the boolean should even simplify the code further.

File: dht-cache/src/cache.rs (hunk 8 above)
Comment (lu-zero) [reply dht-cache/src/cache.rs:255]: We can try that way, I have a pending patch that moves the writes in a separate task using a channel, but if we can get away with a LocaSet it sounds overall cleaner.

File: dht-cache/src/cache.rs (hunk 10 above)
Comment (lu-zero) [reply dht-cache/src/cache.rs:405]: You are absolutely right.

File: dht-cache/src/cache/local.rs [hunk 12]
```diff
@@ -0,0 +1,320 @@
...
+
+    /// Feeds a slice of this type into the given [`Hasher`].
+    pub async fn hash<H: Hasher>(&self, state: &mut H) {
+        let cache = &self.0.read().await.mem;
+        for (topic_name, map_topic_name) in cache.iter() {
+            topic_name.hash(state);
+
+            for (topic_uuid, value) in map_topic_name.iter() {
+                topic_uuid.hash(state);
+                value.to_string().hash(state);
```
[the remark this answers is not in the document]
Comment (dodomorandi) [reply dht-cache/src/cache/local.rs:75]: Is there a way to easily split these comments into a separated issue? Obviously, for me it is easy to _just check the diff_, but I agree that the aim here is a bit different. If we can easily split issues that should be fixed in different PRs, it would be great.

[the remark this answers is not in the document]
Comment (lu-zero) [reply dht-cache/src/cache/local.rs:75]: Sadly there isn't, it would be a nice feature though.

I'd do a round of performance improvements after we are good with the API rework :)

File: dht-cache/src/cache.rs (hunk 11 above)
Comment (lu-zero) [reply dht-cache/src/cache.rs]: You are right, the stream accesses RwLocks, it should be also documented

**Review by dodomorandi — CHANGES_REQUESTED** [review]
Ok, just a major concern, most of the things are minor or considerations.

File: dht-cache/src/cache/local.rs [hunk 13]
```diff
@@ -0,0 +1,320 @@
...
+            .and_modify(|topic| {
+                topic.insert(topic_uuid.to_owned(), elem.to_owned());
+            })
+            .or_insert_with(|| [(topic_uuid.to_owned(), elem.to_owned())].into());
+    }
+}
+
+/// Local cache
+#[derive(Default, Clone)]
+pub struct LocalCache(Arc<RwLock<InnerCache>>);
```
[the remark this answers is not in the document]
Comment (lu-zero) [reply dht-cache/src/cache/local.rs:37]: I started with simply moving the db interaction to a separate task, then we can see if having a localset can integrate with the rest of the system.

File: dht-cache/src/cache/local.rs [hunk 14]
```diff
@@ -0,0 +1,320 @@
...
+pub(crate) struct InnerCache {
+    pub mem: BTreeMap<String, BTreeMap<String, DomoCacheElement>>,
+    pub store: Option<SqlxStorage>,
+}
+
+/// SAFETY: the SqlxStorage access is only over write()
+unsafe impl std::marker::Sync for InnerCache {}
+
+impl InnerCache {
+    pub fn put(&mut self, elem: &DomoCacheElement) {
```
[the remark this answers is not in the document]
Comment (lu-zero) [reply dht-cache/src/cache/local.rs:22]: We need to make the whole api uniform regarding taking reference or taking ownership, ergonomy-wise taking the reference feels nicer.

[the remark this answers is not in the document]
Comment (dodomorandi) [reply dht-cache/src/cache/local.rs:22]: I don't agree, but do what you think it is the best for the project :relaxed:.

File: dht-cache/src/dht.rs [hunk 15]
```diff
@@ -0,0 +1,369 @@
...
+                    }
+                    Event::Ready(peers) => {
+                        log::info!("peers ready: {peers:?}");
+                    }
+                }
+            }
+        }
+
+        drop(b_s);
+        drop(c_s);
```
[the remark this answers is not in the document]
Comment (lu-zero) [reply dht-cache/src/dht.rs:367]: It is now :)

File: dht-cache/src/dht.rs [hunk 16]
```diff
@@ -0,0 +1,390 @@
...
+            log::trace!("Looping {}", swarm.local_peer_id());
+            tokio::select! {
+                // the mdns event is not enough to ensure we can send messages
+                _ = interval.tick() => {
+                    log::debug!("{} Checking for peers", swarm.local_peer_id());
+                    let peers: Vec<_> = check_peers(&mut swarm);
+                        if !peers.is_empty() &&
+                        ev_send.send(Event::Ready(peers)).is_err() {
+                            return swarm;
+                    }
```
Comment (dodomorandi) [inline dht-cache/src/dht.rs:199]: Formatting:
```suggestion
                    let peers: Vec<_> = check_peers(&mut swarm);
                    if !peers.is_empty() && ev_send.send(Event::Ready(peers)).is_err() {
                        return swarm;
                    }
```

**Review by dodomorandi — APPROVED** [review]
